src/v3/plugins/github/edges.js

```diff
@@ -0,0 +1,216 @@
+// @flow
+
+import {
+  type Edge,
+  type EdgeAddressT,
+  type NodeAddressT,
+  EdgeAddress,
+  NodeAddress,
+} from "../../core/graph";
+import * as GithubNode from "./nodes";
+
+export opaque type RawAddress: EdgeAddressT = EdgeAddressT;
+
+// This particular union of types is used frequently. It seems
+// reasonable that it semantically represents some kind of "post" that
+// has a parent, was authored, and can be referred to. But if one of
+// those axes of symmetry breaks down, then we shouldn't feel hesitant
+// to inline this type in the appropriate places.
+export type PostAddress =
+  | GithubNode.CommentAddress
+  | GithubNode.IssueAddress
+  | GithubNode.PullAddress
+  | GithubNode.ReviewAddress;
+
+export type AuthorsAddress = {|
+  +type: "AUTHORS",
+  +author: GithubNode.UserlikeAddress,
+  +content: PostAddress,
+|};
+export type MergedAsAddress = {|
+  +type: "MERGED_AS",
+  +pull: GithubNode.PullAddress,
+|};
+export type HasParentAddress = {|
+  +type: "HAS_PARENT",
+  +child: PostAddress,
+|};
+export type ReferencesAddress = {|
+  +type: "REFERENCES",
+  +referrer: PostAddress,
+  +referent: PostAddress,
+|};
+
+export type StructuredAddress =
+  | AuthorsAddress
+  | MergedAsAddress
+  | HasParentAddress
+  | ReferencesAddress;
+
+export const createEdge = Object.freeze({
+  authors: (
+    author: GithubNode.UserlikeAddress,
+    content: PostAddress
+  ): Edge => ({
+    address: toRaw({type: "AUTHORS", author, content}),
+    src: GithubNode.toRaw(author),
+    dst: GithubNode.toRaw(content),
+  }),
+  mergedAs: (
+    pull: GithubNode.PullAddress,
+    commitAddress: NodeAddressT /* TODO: Make this a Git commit node address. */
+  ): Edge => ({
+    address: toRaw({type: "MERGED_AS", pull}),
+    src: GithubNode.toRaw(pull),
+    dst: commitAddress,
+  }),
+  hasParent: (
+    child: PostAddress,
+    parentAddress: GithubNode.RawAddress
```

Comment:
Why does this take a RawAddress?
Let's define a ParentAddress union alongside PostAddress, and it is a union of RepoAddress | IssueAddress | PullAddress | ReviewAddress.

I believe this will be useful because I already defined that union in my createGraph branch because I also needed it (in addition to using the PostAddress union).

In any case, breaking consistency and requiring a RawAddress for this arg only seems clearly wrong.

Reply:
Okay; it’s fine with me for this to take a structured address. I’m
willing to define and export the `ParentAddress` union, while noting
that it seems pretty weak: while we effectively have `ParentAddress` and
`ChildAddress` types, almost everything is a parent and almost
everything is a child, and we are failing to express (at the type level)
the relationship that (e.g.) a repository cannot be the parent of a
comment. But that seems like a reasonable concession to me.